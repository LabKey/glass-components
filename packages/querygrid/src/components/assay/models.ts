/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { List, OrderedMap, Map, Record } from 'immutable'
import { AppURL, AssayDefinitionModel, AssayUploadTabs, QueryColumn, QueryInfo, FileAttachmentFormModel } from "@glass/base";

export interface AssayPropertiesPanelProps {
    model: AssayWizardModel
    onChange: Function
}

export interface IAssayURLContext {
    assayRequest?: string
    location?: string
    protocol: string
    provider: string
    runId?: string
}

export class AssayWizardModel extends Record({
    assayDef: undefined,
    isError: undefined,
    isWarning: false,
    isInit: false,
    isLoading: false,
    batchId: undefined, // batchId is null for first run
    lastRunId: undefined,
    returnURL: undefined,
    isSubmitted: undefined,
    isSubmitting: undefined,
    errorMsg: undefined,

    attachedFiles: Map<string, File>(),

    batchColumns: OrderedMap<string, QueryColumn>(),
    batchProperties: Map<string, any>(),
    runColumns: OrderedMap<string, QueryColumn>(),
    runId: undefined,
    runProperties: Map<string, any>(),
    runName: undefined,
    comment: undefined,
    dataText: undefined,
    queryInfo: QueryInfo,
    toDelete: undefined,
    selectedSamples: undefined,
}) implements FileAttachmentFormModel {
    assayDef: AssayDefinitionModel;
    isError?: boolean;
    isInit: boolean;
    isLoading: boolean;
    isWarning?: boolean;
    batchId? : number;
    lastRunId?: number;
    returnURL?: AppURL;
    isSubmitted?: boolean;
    isSubmitting?: boolean;
    errorMsg?: string;

    attachedFiles: Map<string, File>;

    batchColumns: OrderedMap<string, QueryColumn>;
    batchProperties: Map<string, any>;
    runColumns: OrderedMap<string, QueryColumn>;
    runId?: string;
    runProperties?: Map<string, any>;
    runName?: string;
    comment?: string;

    dataText: string;
    queryInfo: QueryInfo;
    toDelete?: string;
    selectedSamples?: Map<string, any>;

    constructor(values?: {[key:string]: any}) {
        super(values);
    }

    // TODO shouldn't this move to actions.ts? or put just in Biologics somewhere
    static parseURLContext(location: any, params: any): IAssayURLContext {
        const { protocol, provider } = params;

        let context: IAssayURLContext = {
            location,
            protocol,
            provider
        };

        if (location && location.query) {
            const { assayRequest, runId } = location.query;
            context.assayRequest = assayRequest;
            context.runId = runId;
        }

        return context;
    }

    getAttachedFiles(): List<File> {
        return this.attachedFiles.valueSeq().toList();
    }

    getRunName(currentStep: AssayUploadTabs): string {
        if (this.runName) {
            return this.runName;
        }

        if (currentStep === AssayUploadTabs.Files) {
            // using file upload tab
            const file = this.getAttachedFiles().first();
            if (file) {
                return file.name;
            }
        }

        const date = new Date();
        const dateStr = date.toISOString().split('T')[0];
        let timeStr = date.toTimeString().split(' ')[0];
        timeStr = timeStr.replace(/:/g, '-');
        return this.assayDef.name + '_' + dateStr + '_' + timeStr;
    }
}