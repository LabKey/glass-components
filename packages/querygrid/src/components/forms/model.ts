/*
 * Copyright (c) 2016-2018 LabKey Corporation. All rights reserved. No portion of this work may be reproduced in
 * any form or by any electronic or mechanical means without written permission from LabKey Corporation.
 */
import { List, Map, Record } from 'immutable'
import { Filter } from '@labkey/api'
import { QueryInfo, SchemaQuery } from '@glass/base'

import { ISelectRowsResult } from '../../query/api'
import { DELIMITER } from './input/SelectInput'
import * as actions from './actions'

// This is the same as ReactSelect Option
export interface ReactSelectOption {
    label: string
    value: any
}

export interface QuerySelectModelProps {
    allResults: Map<string, Map<string, any>>
    displayColumn: string
    delimiter: string
    id: string
    isInit: boolean
    maxRows: number
    multiple: boolean
    preLoad: boolean
    queryFilters: List<Filter.IFilter>
    queryInfo: QueryInfo
    rawSelectedValue: any
    schemaQuery: SchemaQuery
    searchResults: Map<string, Map<string, any>>
    selectedQuery: string
    selectedItems: Map<string, any>
    valueColumn: string
}

export class QuerySelectModel extends Record({
    addExactFilter: true,
    allResults: Map<string, Map<string, any>>(),
    displayColumn: undefined,
    delimiter: DELIMITER,
    id: undefined,
    isInit: false,
    maxRows: 20,
    multiple: false,
    preLoad: false,
    queryFilters: undefined,
    queryInfo: undefined,
    rawSelectedValue: undefined,
    schemaQuery: undefined,
    searchResults: Map<string, Map<string, any>>(),
    selectedQuery: '',
    selectedItems: Map<string, any>(),
    valueColumn: undefined
}) implements QuerySelectModelProps {
    addExactFilter: boolean;
    allResults: Map<string, Map<string, any>>;
    displayColumn: string;
    delimiter: string;
    id: string;
    isInit: boolean;
    maxRows: number;
    multiple: boolean;
    preLoad: boolean;
    queryFilters: List<Filter.IFilter>;
    queryInfo: QueryInfo;
    rawSelectedValue: any;
    schemaQuery: SchemaQuery;
    searchResults: Map<string, Map<string, any>>;
    selectedQuery: string;
    selectedItems: Map<string, any>;
    valueColumn: string;

    constructor(values?: Partial<QuerySelectModelProps>) {
        super(values);
    }

    formatSavedResults(data?: Map<string, Map<string, any>>, token?: string): Array<ReactSelectOption> {
        return actions.formatSavedResults(this, data, token);
    }

    getSelectedOptions(): ReactSelectOption | Array<ReactSelectOption> {
        const options = actions.formatResults(this, this.selectedItems);

        if (this.multiple) {
            return options;
        }
        else if (options.length === 1) {
            return options[0];
        }
        else if (options.length > 1) {
            console.warn('QuerySelect.getSelectedOptions: There are more than one options available, however, must in in "multiple" mode.');
        }

        return undefined;
    }

    parseSearch(input: any): boolean | string {
        return actions.selectShouldSearch(this, input);
    }

    saveSearchResults(data: Map<string, Map<string, any>>) {
        return actions.saveSearchResults(this, data);
    }

    setSelection(value: any) {
        return actions.setSelection(this, value);
    }

    search(input: any): Promise<ISelectRowsResult> {
        return actions.fetchSearchResults(this, input);
    }
}

export interface ISelectInitData {
    name: string
    type: string
    getSelectComponentId(): string
}