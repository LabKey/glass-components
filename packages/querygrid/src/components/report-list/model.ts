/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { AppURL } from '@glass/base';
import { IDataViewInfo } from '../../models';

export type ReportURLMapper = (report: IDataViewInfo) => AppURL

/**
 * FlattenResponse converts the repsonse body (a nested tree structure) from browseDataTree into a flat list of
 * ReportItem objects. This method purposely ignores categories and their nested structures.
 *
 * @param response: the body from the browseDataTree API Action
 * @param urlMapper: ReportURLMapper
 */
export function flattenBrowseDataTreeResponse(response: any, urlMapper: ReportURLMapper): Array<IDataViewInfo> {
    function _flattenBrowseDataTreeResponse(all, item): Array<IDataViewInfo> {
        if (item.hasOwnProperty('children')) {
            return [...all, ...item.children.reduce(_flattenBrowseDataTreeResponse, [])] as Array<IDataViewInfo>;
        } else {
            const appUrl = urlMapper(item);

            if (appUrl !== item.runUrl) {
                item.appUrl = appUrl;
            }

            return [...all, item];
        }
    }

    return _flattenBrowseDataTreeResponse([], response);
}
