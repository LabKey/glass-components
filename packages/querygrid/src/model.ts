/*
 * Copyright (c) 2019 LabKey Corporation. All rights reserved. No portion of this work may be reproduced in
 * any form or by any electronic or mechanical means without written permission from LabKey Corporation.
 */
import { List, Map, OrderedMap, Record, Set } from 'immutable'
import { Filter } from '@labkey/api'
import {
    QueryColumn, QueryInfo, QueryGridModel, SchemaQuery, ViewInfo,
    IQueryGridModel, IGridLoader, resolveSchemaQuery
} from '@glass/models'

import { genCellKey } from './actions'
import { getQueryGridModel } from './global'
import { DefaultGridLoader } from './components/GridLoader'

const emptyList = List<string>();

interface IStateModelProps {
    allowSelection?: boolean
    baseFilters?: List<Filter.Filter>
    bindURL?: boolean
    editable?: boolean
    isPaged?: boolean
    loader?: IGridLoader
    maxRows?: number
    queryInfo?: QueryInfo
    requiredColumns?: List<string>
    sorts?: string
    sortable?: boolean
    title?: string
    urlPrefix?: string
    omittedColumns?: List<string>
}

export function getStateModelId(gridId: string, schemaQuery: SchemaQuery, keyValue?: any): string {
    let parts = [gridId, resolveSchemaQuery(schemaQuery)];

    if (schemaQuery && schemaQuery.viewName) {
        parts.push(schemaQuery.viewName);
    }
    if (keyValue !== undefined) {
        parts.push(keyValue);
    }

    return parts.join('|').toLowerCase();
}

/**
 * Used to create a QueryGridModel, based on some initial props, that can be put into the global state.
 * @param gridId
 * @param schemaQuery
 * @param [initProps] can be either a props object or a function that returns a props object. The advantage of using
 * a function is that it is only called once for the lifetime of the model thus saving cycles constructing the prop
 * object.
 * @param [keyValue]
 * @returns {QueryGridModel}
 */
export function getStateQueryGridModel(
    gridId: string,
    schemaQuery: SchemaQuery,
    initProps?: IStateModelProps | Function, // () => IStateModelProps
    keyValue?: any
): QueryGridModel {
    const modelId = getStateModelId(gridId, schemaQuery, keyValue);

    // if the model already exists in the global state, return it
    const model = getQueryGridModel(modelId);
    if (model) {
        return model;
    }

    let modelProps: Partial<IQueryGridModel> = {
        allowSelection: true,
        baseFilters: List<Filter.Filter>(),
        bindURL: true,
        editable: false,
        id: modelId,
        isPaged: false, // Figure out how to set this to the same default value as the model
        loader: DefaultGridLoader,
        keyValue: undefined,
        maxRows: 20,
        schema: schemaQuery.schemaName,
        query: schemaQuery.queryName,
        queryInfo: undefined,
        requiredColumns: emptyList,
        sorts: undefined,
        sortable: true,
        title: undefined,
        urlPrefix: undefined,
        view: schemaQuery.viewName,
        omittedColumns: emptyList
    };

    if (keyValue !== undefined) {
        modelProps.keyValue = keyValue;

        if (schemaQuery.viewName === undefined) {
            modelProps.view = ViewInfo.DETAIL_NAME;
            modelProps.bindURL = false;
        }
    }

    let props: IStateModelProps;
    if (initProps !== undefined) {
        props = typeof initProps === 'function' ? initProps() : initProps;

        if (props) {
            if (props.bindURL !== undefined) {
                modelProps.bindURL = props.bindURL === true;
            }

            if (props.isPaged !== undefined) {
                modelProps.isPaged = props.isPaged === true;
            }

            if (props.loader !== undefined) {
                modelProps.loader = props.loader;
            }

            if (props.queryInfo !== undefined) {
                modelProps.queryInfo = props.queryInfo;
            }

            if (props.maxRows !== undefined) {
                modelProps.maxRows = props.maxRows;
            }

            if (props.baseFilters) {
                modelProps.baseFilters = props.baseFilters;
            }

            if (props.requiredColumns !== undefined) {
                modelProps.requiredColumns = props.requiredColumns;
            }
            if (props.urlPrefix !== undefined) {
                modelProps.urlPrefix = props.urlPrefix;
            }

            if (props.title !== undefined) {
                modelProps.title = props.title;
            }

            if (props.allowSelection !== undefined) {
                modelProps.allowSelection = props.allowSelection;
            }

            if (props.editable !== undefined) {
                modelProps.editable = props.editable;
            }

            if (props.sortable !== undefined) {
                modelProps.sortable = props.sortable;
            }

            if (props.sorts !== undefined) {
                modelProps.sorts = props.sorts;
            }

            if (props.omittedColumns !== undefined) {
                modelProps.omittedColumns = props.omittedColumns;
            }
        }
    }

    return new QueryGridModel(modelProps);
}

// commented out attributes are not used in app
export class DataViewInfo extends Record({
    name: undefined,
    type: undefined,
    reportId: undefined,
    schemaName: undefined,
    queryName: undefined,
    shared: false,
    iconCls: undefined,
    description: undefined,
    // container: undefined,
    // access: undefined,
    // detailsUrl: undefined,
    // icon: undefined,
    // runUrl: undefined,
    // defaultThumbnailUrl: undefined,
    // defaultIconCls: undefined,
    // modified: undefined,
    // modifiedBy: undefined,
    // id: undefined,
    // cratedByuserId: undefined,
    // showInDashboard: undefined,
    // thumbnail: undefined,
    // visible: undefined,
    // contentModified: undefined,
    // author: undefined,
    // created: undefined,
    // dataType: undefined,
    // readOnly: undefined,
    // createdBy: undefined,
    // allowCustomThumbnail: undefined,
    // category: undefined,

    // our stuff
    isLoading: false,
    isLoaded: false,
    error: undefined
}) {
    name: string;
    type: string;
    reportId: string;
    schemaName: string;
    queryName: string;
    shared: boolean;
    iconCls: string;
    description: string;
    // container: string;
    // access: any;
    // detailsUrl: string;
    // icon: string;
    // runUrl: string;
    // defaultThumbnailUrl: string;
    // defaultIconCls: string;
    // modified: date;
    // modifiedBy: number;
    // id: string;
    // cratedByuserId: number;
    // showInDashboard: boolean;
    // thumbnail: string;
    // visible: boolean;
    // contentModified: date;
    // author: string;
    // created: date;
    // dataType: string;
    // readOnly: boolean;
    // createdBy: number;
    // allowCustomThumbnail: boolean;
    // category: any;
    isLoading: boolean;
    isLoaded: boolean;
    error: string;

    constructor(values?: {[key:string]: any}) {
        super(values);
    }

    getLabel() {
        return this.name;
    }

    isShared() {
        return this.shared;
    }

    getIconCls() {
        return this.iconCls;
    }

    isVisChartType() {
        return this.type === "Bar Chart"
            || this.type === "Box and Whisker Plot"
            || this.type === "Pie Chart"
            || this.type === "XY Scatter Plot"
            || this.type === "XY Series Line Plot";
    }
}

export class VisualizationConfigModel extends Record({
    queryConfig: undefined,
    chartConfig: undefined
}) {
    queryConfig: QueryConfigModel;
    chartConfig: ChartConfigModel;

    static create(raw: any): VisualizationConfigModel {
        return new VisualizationConfigModel(Object.assign({}, raw, {
            chartConfig: new ChartConfigModel(raw.chartConfig),
            queryConfig: new QueryConfigModel(raw.queryConfig)
        }))
    }

    constructor(values?: {[key:string]: any}) {
        super(values);
    }
}

export class ChartConfigModel extends Record({
    geomOptions: undefined,
    height: undefined,
    labels: undefined,
    measures: undefined,
    pointType: undefined,
    renderType: undefined,
    scales: undefined,
    width: undefined
}) {
    geomOptions: any;
    height: number;
    labels: any;
    measures: any;
    pointType: string;
    renderType: string;
    scales: any;
    width: number;

    constructor(values?: {[key:string]: any}) {
        super(values);
    }
}

export class QueryConfigModel extends Record({
    columns: undefined,
    containerPath: undefined,
    // dataRegionName: undefined,
    filterArray: undefined,
    maxRows: undefined,
    method: undefined,
    parameters: undefined,
    // queryLabel: undefined,
    queryName: undefined,
    requiredVersion: undefined,
    schemaName: undefined,
    // sort: undefined,
    viewName: undefined
}) {
    columns: List<string>;
    containerPath: string;
    // dataRegionName: string;
    filterArray: List<any>;
    maxRows: number;
    method: string;
    parameters: any;
    // queryLabel: string;
    queryName: string;
    requiredVersion: string;
    schemaName: string;
    // sort: string;
    viewName: string;

    constructor(values?: {[key:string]: any}) {
        super(values);
    }
}

export interface ValueDescriptor {
    display: any
    raw: any
}

export interface CellMessage {
    message: string
}

export type CellMessages = Map<string, CellMessage>;
export type CellValues = Map<string, List<ValueDescriptor>>;

export interface EditorModelProps {
    cellMessages: CellMessages
    cellValues: CellValues
    colCount: number
    id: string
    isPasting: boolean
    focusColIdx: number
    focusRowIdx: number
    focusValue: List<ValueDescriptor>
    numPastedRows: number
    rowCount: number
    selectedColIdx: number
    selectedRowIdx: number
    selectionCells: Set<string>
}

export class EditorModel extends Record({
    cellMessages: Map<string, CellMessage>(),
    cellValues: Map<string, List<ValueDescriptor>>(),
    colCount: 0,
    id: undefined,
    isPasting: false,
    focusColIdx: -1,
    focusRowIdx: -1,
    focusValue: undefined,
    numPastedRows: 0,
    rowCount: 0,
    selectedColIdx: -1,
    selectedRowIdx: -1,
    selectionCells: Set<string>()
}) implements EditorModelProps {
    cellMessages: CellMessages;
    cellValues: CellValues;
    colCount: number;
    id: string;
    isPasting: boolean;
    focusColIdx: number;
    focusRowIdx: number;
    focusValue: List<ValueDescriptor>;
    numPastedRows: number;
    rowCount: number;
    selectedColIdx: number;
    selectedRowIdx: number;
    selectionCells: Set<string>;

    constructor(values?: {[key:string]: any}) {
        super(values);
    }

    findNextCell(startCol: number, startRow: number,
                 predicate: (value: List<ValueDescriptor>, colIdx: number, rowIdx: number) => boolean,
                 advance: (colIdx: number, rowIdx: number) => {colIdx: number, rowIdx: number})
    {
        let colIdx = startCol,
            rowIdx = startRow;

        while (true) {
            ({colIdx, rowIdx} = advance(colIdx, rowIdx));
            if (!this.isInBounds(colIdx, rowIdx))
                break;

            let value = this.getValue(colIdx, rowIdx);
            if (predicate(value, colIdx, rowIdx)) {
                return {
                    value,
                    colIdx,
                    rowIdx
                };
            }
        }

        // not found
        return null;
    }

    getMessage(colIdx: number, rowIdx: number): CellMessage {
        return this.cellMessages.get(genCellKey(colIdx, rowIdx));
    }

    getRawData(model: QueryGridModel): List<Map<string, any>> {
        let data = List<Map<string, any>>().asMutable();
        const columns = model.getInsertColumns();

        for (let rn = 0; rn < model.data.size; rn++) {
            let row = Map<string, any>().asMutable();
            columns.forEach((col, cn) => {
                const values = this.getValue(cn, rn);

                if (col.isLookup()) {
                    if (col.isExpInput()) {
                        let sep = '';
                        row.set(col.name, values.reduce((str, vd) => {
                            if (vd.display !== undefined && vd.display !== null) {
                                str += sep + vd.display;
                                sep = ', ';
                            }
                            return str;
                        }, ''));
                        return;
                    }
                    else if (col.isJunctionLookup()) {
                        row.set(col.name, values.reduce((arr, vd) => {
                            if (vd.raw !== undefined && vd.raw !== null) {
                                arr.push(vd.raw);
                            }
                            return arr;
                        }, []));
                        return;
                    }
                }

                row.set(col.name, values.size === 1 ? values.first().raw : undefined);
            });

            data.push(row.asImmutable());
        }

        return data.asImmutable();
    }

    getValue(colIdx: number, rowIdx: number): List<ValueDescriptor> {
        const cellKey = genCellKey(colIdx, rowIdx);
        if (this.cellValues.has(cellKey)) {
            return this.cellValues.get(cellKey);
        }

        return List<ValueDescriptor>();
    }

    hasFocus(): boolean {
        return this.focusColIdx > -1 && this.focusRowIdx > -1;
    }

    hasMultipleSelection(): boolean {
        return this.selectionCells.size > 1;
    }

    hasSelection(): boolean {
        return this.selectedColIdx > -1 && this.selectedRowIdx > -1;
    }

    isInBounds(colIdx: number, rowIdx: number): boolean {
        return (colIdx >= 0 && colIdx < this.colCount && rowIdx >= 0 && rowIdx < this.rowCount);
    }

    inSelection(colIdx: number, rowIdx: number): boolean {
        return (
            colIdx > -1 && rowIdx > -1 &&
            this.selectionCells.get(genCellKey(colIdx, rowIdx)) !== undefined
        )
    }

    isFocused(colIdx: number, rowIdx: number): boolean {
        return (
            colIdx > -1 && rowIdx > -1 &&
            this.focusColIdx === colIdx &&
            this.focusRowIdx === rowIdx
        );
    }

    isSelected(colIdx: number, rowIdx: number): boolean {
        return (
            colIdx > -1 && rowIdx > -1 &&
            this.selectedColIdx === colIdx &&
            this.selectedRowIdx === rowIdx
        );
    }
}

export class LookupStore extends Record({
    key: undefined,
    descriptors: OrderedMap<any, ValueDescriptor>(),
    isLoaded: false,
    isLoading: false,
    lastToken: '~~INITIAL_TOKEN~~',
    loadCount: 0,
    matchCount: 0
}) {
    key: string;
    descriptors: OrderedMap<any, ValueDescriptor>;
    isLoaded: boolean;
    isLoading: boolean;
    lastToken: string;
    loadCount: number;
    matchCount: number;

    static key(col: QueryColumn): string {
        return [
            col.lookup.schemaName,
            col.lookup.queryName,
            col.fieldKey
        ].join('|');
    }

    constructor(values?: {[key:string]: any}) {
        super(values);
    }

    containsAll(values: List<string>): boolean {
        let displayValues = this.descriptors
            .reduce((valueSet, descriptor) => valueSet.add(descriptor.display), Set<String>());

        return values.find((value) => !displayValues.contains(value)) == undefined;
    }
}