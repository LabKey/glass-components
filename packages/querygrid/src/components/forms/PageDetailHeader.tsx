/*
 * Copyright (c) 2017-2018 LabKey Corporation. All rights reserved. No portion of this work may be reproduced in
 * any form or by any electronic or mechanical means without written permission from LabKey Corporation.
 */
import * as React from 'react'
import { hasAllPermissions, PermissionTypes, SVGIcon, User } from "@glass/base";

// import { SVGIcon } from '../../components/Image/SVGIcon'
import { FieldEditTrigger, FieldEditTriggerProps } from './FieldEditTrigger'

interface Props {
    user: User
    content?: Array<any>
    description?: string
    fieldTriggerProps?: FieldEditTriggerProps
    iconDir?: string
    iconSrc?: string
    iconUrl?: string
    leftColumns?: number
    subTitle?: any
    title: any
}

export class PageDetailHeader extends React.Component<Props, any> {

    static defaultProps = {
        leftColumns: 6
    };

    render() {
        const { children, description, fieldTriggerProps, iconUrl, iconDir, iconSrc, leftColumns, subTitle, title, user } = this.props;

        return (
            <div className="page-header">
                <div className={`col-md-${leftColumns} detail__header--container`}>
                    <div className="detail__header--image-container">
                        {iconUrl ? <img
                                src={iconUrl}
                                className="detail__header-icon"
                            />
                            : <SVGIcon
                                iconDir={iconDir}
                                iconSrc={iconSrc ? iconSrc : ''}
                                className="detail__header-icon"
                            />
                        }
                    </div>
                    <h2 className="no-margin-top detail__header--name">{title}</h2>
                    {subTitle && (
                        <h4 className="test-loc-detail-subtitle">{subTitle}</h4>
                    )}
                    {description && (
                        <span className="detail__header--desc">{description}</span>
                    )}
                    {fieldTriggerProps && (
                        <div className="text__truncate">
                            <FieldEditTrigger {...fieldTriggerProps}
                                canUpdate={hasAllPermissions(user, [PermissionTypes.Update])}
                            />
                        </div>
                    )}
                </div>

                {children && (
                    <div className="pull-right">
                        {children}
                    </div>
                )}

                <div className="clearfix"/>
            </div>
        )
    }
}
