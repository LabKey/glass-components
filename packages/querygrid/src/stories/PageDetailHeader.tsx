import * as React from 'react';
import { Map, fromJS } from 'immutable';
import { storiesOf } from "@storybook/react";
import { boolean, number, text, withKnobs } from '@storybook/addon-knobs'
import { CreatedModified } from "@glass/base";

import { PageDetailHeader } from "../components/forms/PageDetailHeader";
import './stories.scss'

storiesOf('PageDetailHeader', module)
    .addDecorator(withKnobs)
    .add("with knobs", () => {
        const createdRow = Map<string, any>(fromJS({
            Created: {
                formattedValue: "2019-05-15 19:45",
                value: "2019-05-15 19:45:40.593"
            },
            CreatedBy: {
                displayValue: "username",
                url: "#/q/core/siteusers/1001",
                value: 1001
            },
            Modified: {
                formattedValue: "2019-05-16 19:45",
                value: "2019-05-16 19:45:40.593"
            },
            ModifiedBy: {
                displayValue: "username2",
                url: "#/q/core/siteusers/1002",
                value: 1002
            }
        }));

        return (
            <PageDetailHeader
                user={null}
                iconUrl={'http://labkey.wpengine.com/wp-content/uploads/2015/12/cropped-LK-icon.png'}
                title={text('title', 'Page Detail Header')}
                subTitle={text('subtitle', 'With a subtitle')}
            >
                <CreatedModified row={createdRow}/>
            </PageDetailHeader>
        )
    });