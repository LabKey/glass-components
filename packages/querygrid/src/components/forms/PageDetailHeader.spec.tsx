/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'

import { PageDetailHeader } from "./PageDetailHeader";
import { User } from '../base/models/model';

describe("<PageDetailHeader/>", () => {

   test("default props", () => {
      const component = (
          <PageDetailHeader title={'Title'} user={new User()}/>
      );

      const tree = renderer.create(component).toJSON();
      expect(tree).toMatchSnapshot();
   });

   test("with additional props", () => {
      const component = (
          <PageDetailHeader
              user={new User()}
              title={'Title'}
              subTitle={'Subtitle'}
              description={'Description'}
              iconDir={'iconDir'}
              iconSrc={'iconSrc'}
              leftColumns={5}
          >
             <div>Someting off to the right</div>
          </PageDetailHeader>
      );

      const tree = renderer.create(component).toJSON();
      expect(tree).toMatchSnapshot();
   });

   test("prefer iconUrl", () => {
      const component = (
          <PageDetailHeader
              user={new User()}
              title={'Title'}
              iconUrl={'iconUrl'}
              iconDir={'iconDir'}
              iconSrc={'iconSrc'}
          />
      );

      const wrapper = mount(component);
      const srcAttr = wrapper.find('img').getDOMNode().getAttribute('src');
      expect(srcAttr).toBe('iconUrl');
      wrapper.unmount();
   });

});