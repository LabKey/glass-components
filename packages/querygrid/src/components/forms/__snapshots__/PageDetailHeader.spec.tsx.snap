// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageDetailHeader/> default props 1`] = `
<div
  className="page-header"
>
  <div
    className="col-md-6 detail__header--container"
  >
    <div
      className="detail__header--image-container"
    >
      <img
        alt="-icon"
        className="detail__header-icon"
        height="100%"
        src="/labkey//images/default.svg"
        width="100%"
      />
    </div>
    <h2
      className="no-margin-top detail__header--name"
    >
      Title
    </h2>
  </div>
  <div
    className="clearfix"
  />
</div>
`;

exports[`<PageDetailHeader/> with additional props 1`] = `
<div
  className="page-header"
>
  <div
    className="col-md-5 detail__header--container"
  >
    <div
      className="detail__header--image-container"
    >
      <img
        alt="iconSrc-icon"
        className="detail__header-icon"
        height="100%"
        src="/labkey/iconDir/images/iconsrc.svg"
        width="100%"
      />
    </div>
    <h2
      className="no-margin-top detail__header--name"
    >
      Title
    </h2>
    <h4
      className="test-loc-detail-subtitle"
    >
      Subtitle
    </h4>
    <span
      className="detail__header--desc"
    >
      Description
    </span>
  </div>
  <div
    className="pull-right"
  >
    <div>
      Someting off to the right
    </div>
  </div>
  <div
    className="clearfix"
  />
</div>
`;
